lxd/storage_volumes_snapshot.go

```diff
@@ -406,32 +429,55 @@ func storagePoolVolumeSnapshotTypePut(d *Daemon, r *http.Request) response.Respo
 		return resp
 	}
 
-	_, vol, err := d.cluster.StoragePoolNodeVolumeGetTypeByProject(projectName, fullSnapshotName, volumeType, poolID)
+	volID, vol, err := d.cluster.StoragePoolNodeVolumeGetTypeByProject(projectName, fullSnapshotName, volumeType, poolID)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	expiry, err := d.cluster.StorageVolumeSnapshotExpiryGet(volID)
 	if err != nil {
 		return response.SmartError(err)
 	}
 
 	// Validate the ETag
-	etag := []interface{}{snapshotName, vol.Description, vol.Config}
+	etag := []interface{}{snapshotName, vol.Description, vol.Config, expiry}
 	err = util.EtagCheck(r, etag)
 	if err != nil {
 		return response.PreconditionFailed(err)
 	}
 
+	rj := shared.Jmap{}
+
+	err = json.NewDecoder(r.Body).Decode(&rj)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
 	req := api.StorageVolumeSnapshotPut{}
```

Comment:
@monstermunchkin I dont quite understand this part, the `StorageVolumeSnapshotPut` now has an `ExpiresAt` property, and previously the request was being decoded into an instance of `StorageVolumeSnapshotPut`, so this should have made the `ExpiresAt` property available. However now the request body is being decoded into a `shared.Jmap{}`. And  then if the body contains the `expires_at` field, then the original req.ExpiresAt field is used. I'm confused :)

Reply:
Using the `shared.Jmap{}`, we can check whether the `expires_at` field is set. This is the only reason we use it here. We then run `Marshal` and `Unmarshal` again, this time filling the `req`. We cannot simply use `r.Body` again since that has already been read before and would return nothing. Anyway, if the field is set, we set `expiry` which overrides the database value.

Does that help with the confusion? :)

Reply:
@monstermunchkin could you make `ExpiresAt` a `*time.Time` like you did in `StorageVolumeSnapshotsPost` that way detecting if it was nil would indicate if it was supplied or not.

Reply:
Yes, I could do that. For some reason, we went for the `shared.Jmap{}` solution for instance snapshots as well. @stgraber do you still remember the reason?

Reply:
I think it boils down to consistency with what we already have.

If we were to redo those APIs we would certainly use a pointer, but we didn't when first implementing them, so while it's safe for us to accept a pointer on Post because LXD can do the proper validation, we don't want to have half of them be pointers and the other half be structs in the data read by our users as someone is bound to miss a `!= nil` check when copy/pasting logic and will get a fun surprise when doing that.